kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,324 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(CONSTRAINT_DATA_H)
+#define CONSTRAINT_DATA_H
+// System includes
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+#include <tuple>
+#include <utility>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h" 
+#include "includes/dof.h"
+#include "includes/node.h"
+
+namespace Kratos
+{
+
+// ConstraintEquation class start
+/*
+*   This class stores the information regarding the constraint equation. Naming convenction is defined like this.
+*   (each object of this class will store one equation in the given form)
+*
+*   slaveDOF = w_1*masterDOF_1 + w_2*masterDOF_2 + ..... + w_n*masterDOF_n
+*   
+*   each slaveDOF and each of its masterDOF have the following attributes 
+*   a. dof ID
+*   b. dof KEY
+*   c. equation ID
+*
+*   one should be able to access the constraint equation both with (dofID , dofKEY) pair and/or equationID of the slave.
+*   This equation is imposed on the linear system of equations either element wise or on the global system.
+*
+*/
+class ConstraintEquation
+{
+  private:
+    typedef Dof<double> DofType;
+    struct MasterData
+    {
+        KRATOS_CLASS_POINTER_DEFINITION(MasterData);
+        MasterData(DofType const &iMasterDof, double iWeight = 0.0) : masterWeight(iWeight), id(iMasterDof.Id()), key(iMasterDof.GetVariable().Key())
+        {
+        }
+        friend bool operator==(const MasterData::Pointer &iMasterDataOne, const DofType &iMasterDof)
+        {
+            return ((iMasterDataOne->MasterDofId() == iMasterDof.Id()) && (iMasterDataOne->MasterKey() == iMasterDof.GetVariable().Key()));
+        }
+        unsigned int MasterKey() { return key; }
+        double MasterWeight() { return masterWeight; }
+        void UpdateWeight(double iWeight) { masterWeight += iWeight; }
+        unsigned int MasterDofId() { return id; }
+        void SetEquationId(unsigned int iId) { equationId = iId; }
+        unsigned int MasterEqId() { return equationId; }
+        void SetMasterEqId(unsigned int iId) { equationId = iId; }
+
+      private:
+        double masterWeight;
+        const unsigned int id;

```

Comment:
@pooyan-dadvand i think we best avoid having dofs since it would be very hard to sync them in mpi. with the key and node index we can get them as we need

Reply:
I tried to keep a reference/pointer to the dof here. But the problem is in the event of reconstruction of dofs the pointer or the ref becomes invalid and this is a problem.

Reply:
I see